### PR TITLE
refactor(distribution-inference): migrate to new attack architecture

### PR DESCRIPTION
**Pull request by asim29**
## Summary

Deletes the 528-line monolithic `DistributionInference` class and replaces it with two focused, independently testable concrete attacks that implement `DistributionInferenceAttack`. Adds a standalone metrics module and full unit + integration test coverage.

**Stacked on:** `pr/dataset-loaders-and-amuletdataset-extensions`

---

### Deleted

| File | Reason |
|------|--------|
| `amulet/distribution_inference/attacks/distribution_inference.py` | 528-line monolith mixing data-loading, filtering, model training, and inference logic. Replaced by two focused classes below. |

#### `attacks/suri_evans_2022.py` — `SuriEvans2022`

KL-divergence distinguishing test from [Suri & Evans, NeurIPS 2022](https://arxiv.org/abs/2109.06024).

- Collects per-model softmax predictions on both test distributions.
- Computes pairwise KL-divergence scores between adversary and victim predictions.
- Returns `{"predictions": ..., "ground_truth": ...}` following the standard `attack()` contract.
- No meta-classifier or weight extraction — purely black-box.

#### `attacks/white_box_pim.py` — `WhiteBoxPIM`

White-box attack using Permutation Invariant Models (PIM), also from Suri & Evans 2022.

- Extracts raw layer weights from adversary model populations via `utils/__weight_extraction.get_layer_parameters`.
- Trains a `PermInvModel` meta-classifier (from `utils/__meta_classifiers`) to distinguish distribution 1 from distribution 2.
- Evaluates the trained meta-classifier on held-out victim models.
- This attack is the primary consumer of the `__meta_classifiers` and `__weight_extraction` utilities landed in PR 1.

Both attacks inherit the full model-population lifecycle (`train_model_population`, `prepare_model_populations`) from `DistributionInferenceAttack` (landed in PR 2).

### New metrics module

`metrics/distinguishing_accuracy.py` — `evaluate_distinguishing_accuracy(predictions, ground_truth, threshold)`:
- Returns `distinguishing_accuracy` (fraction correct at threshold) and `auc_score` (ROC-AUC).
- Decoupled from any attack class — takes plain numpy arrays.

### `attacks/__init__.py`

- Removes `DistributionInference` export.
- Adds `SuriEvans2022` and `WhiteBoxPIM`.

### Examples (pulled in to fix `basedpyright` failure)

`run_distribution_inference.py` previously imported the now-deleted `DistributionInference`. Updated to use `SuriEvans2022`. `run_whitebox_dist_inference.py` is a new runnable pipeline for the white-box attack. Both are full PR 5 changes but were required to keep CI green on this branch.

### Tests

| File | Type | What it covers |
|------|------|---------------|
| `tests/unit/test_metrics_distribution.py` | unit | `evaluate_distinguishing_accuracy` edge cases |
| `tests/unit/test_distribution_dataset_utils.py` | unit | `DistributionSplits` and `prepare_distribution_splits` |
| `tests/unit/test_distribution_inference_attack_base.py` | unit | `DistributionInferenceAttack` lifecycle |
| `tests/integration/test_whitebox_pim.py` | integration | End-to-end `WhiteBoxPIM` run on Census |

---

## Test plan

- [ ] `uv run pre-commit run --all-files` passes
- [ ] `uv run pytest tests/unit/test_metrics_distribution.py tests/unit/test_distribution_dataset_utils.py tests/unit/test_distribution_inference_attack_base.py -v` passes
- [ ] `from amulet.distribution_inference.attacks import SuriEvans2022, WhiteBoxPIM` resolves
- [ ] `from amulet.distribution_inference.metrics import evaluate_distinguishing_accuracy` resolves
- [ ] `from amulet.distribution_inference.attacks import DistributionInference` raises `ImportError`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: amulet/distribution_inference/attacks/suri_evans_2022.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+        return np.stack(per_model_preds, axis=0), ground_truth
+
+    @staticmethod
+    def _sigmoid(x: np.ndarray) -> np.ndarray:
+        exp = np.exp(x)
+        return exp / (1 + exp)
+
+    @staticmethod
+    def _kl_divergence(x: np.ndarray, y: np.ndarray) -> np.ndarray:
```
Comment (sebszyller) [inline amulet/distribution_inference/attacks/suri_evans_2022.py]: why do we need to reimplement KL? it's in pytorch

**Review by sebszyller — CHANGES_REQUESTED** [review]

File: amulet/distribution_inference/attacks/suri_evans_2022.py (hunk 1 above)
Comment (asim29) [reply amulet/distribution_inference/attacks/suri_evans_2022.py]: fixed in [4ea78d5](https://github.com/ssg-research/amulet/commit/4ea78d52cc30fb7338bd65a7f30a100a35ba3599) and [d5f6ff6](https://github.com/ssg-research/amulet/pull/74/commits/d5f6ff60d89274aaf7b69c6be5ba8df75a6c5aae)

**Review by sebszyller — APPROVED** [review]
